File: packages/contracts/src/m2la_contracts/validate.py
"""Validation report contracts."""

from pydantic import BaseModel, Field

from m2la_contracts.enums import Severity
from m2la_contracts.telemetry import TelemetryContext


class ValidationIssue(BaseModel):
    """A single validation issue found in the generated artifacts."""

    rule_id: str = Field(..., description="Machine-readable rule identifier (e.g., 'SCHEMA_001')")
    message: str = Field(..., description="Human-readable description of the issue")
    severity: Severity = Field(..., description="Issue severity")
    artifact_path: str | None = Field(default=None, description="Path to the artifact containing the issue")
    location: str | None = Field(
        default=None, description="Location within the artifact (e.g., line number or JSON path)"
    )


class ValidationReport(BaseModel):
    """Report from validating generated Logic Apps artifacts."""

    valid: bool = Field(..., description="Whether all artifacts passed validation")
    issues: list[ValidationIssue] = Field(default_factory=list, description="List of validation issues found")
    artifacts_validated: int = Field(default=0, ge=0, description="Number of artifacts that were validated")
    telemetry: TelemetryContext = Field(..., description="Telemetry context for trace correlation")

File: packages/contracts/typescript/src/validate.ts
/**
 * Validation request/report contracts.
 *
 * Note: The validate request is sent as multipart/form-data (file upload),
 * not as a JSON body.  The ValidateRequest type below describes the
 * logical fields; the actual upload is handled by the CLI's api-client.
 */

import type { Severity } from "./enums.js";
import type { TelemetryContext } from "./telemetry.js";

/**
 * Logical request to validate generated Logic Apps artifacts.
 *
 * In practice, the output artifacts are uploaded as a zip file via
 * multipart/form-data.  The telemetry is sent as a form field.
 */
export interface ValidateRequest {
  /** Telemetry context for trace propagation. */
  telemetry?: TelemetryContext | null;
}

/** A single validation issue found in the generated artifacts. */
export interface ValidationIssue {
  /** Machine-readable rule identifier (e.g., 'SCHEMA_001'). */
  rule_id: string;
  /** Human-readable description of the issue. */
  message: string;
  /** Issue severity. */
  severity: Severity;
  /** Path to the artifact containing the issue. */
  artifact_path?: string | null;
  /** Location within the artifact (e.g., line number or JSON path). */
  location?: string | null;
}

/** Report from validating generated Logic Apps artifacts. */
export interface ValidationReport {
  /** Whether all artifacts passed validation. */
  valid: boolean;
  /** List of validation issues found. */
  issues: ValidationIssue[];
  /** Number of artifacts that were validated. */
  artifacts_validated: number;
  /** Telemetry context for trace correlation. */
  telemetry: TelemetryContext;
}
